tools/autograd/templates/variable_factories.h

```diff
@@ -22,61 +22,96 @@ using at::DimnameList;
 namespace torch {
 
 namespace detail {
-  enum class InitListTensorType { Scalar, InitList };
+  enum class ListInitTensorType { Scalar, Vector };
 
-  // We use `InitListTensor` to support converting an arbitrarily nested braced-init-list
-  // (e.g. {{1, 2}, {3, 4}}) into the equivalent Tensor, taking advantage of the fact that
+  // `ListInitTensor` represents a Tensor initialized by a list type or a scalar type
+  // (which is convertible to a list type). Specifically, it supports converting the
+  // following types to a Tensor:
+  // - Scalar value, e.g. `3.14`.
+  // - `at::ArrayRef<T>`, e.g. `at::ArrayRef<int>({1, 2, 3})`.
+  // - `std::vector<T>`, e.g. `std::vector<int>({1, 2, 3})`.
+  // - Arbitrarily nested braced-init-list (e.g. {{1, 2}, {3, 4}}), taking advantage of the fact that
   // the constructor will automatically be called recursively until it reaches all innermost
   // scalar values.
   //
-  // At any time, a `InitListTensor` object represents either of the following:
+  // At any time, a `ListInitTensor` object represents either of the following:
   // 1. A scalar with value `scalar()` and type `scalar_type()`.
-  // 2. A Tensor represented in `std::initializer_list<InitListTensor>` form, with value
-  //    `init_list()`, Tensor scalar type `scalar_type()`, and Tensor sizes `sizes()`.
-  struct InitListTensor {
-#define TENSOR(T, S)                   \
-    InitListTensor(T scalar) :         \
-        scalar_(scalar), init_list_(), \
-        sizes_(),                      \
-        scalar_type_(at::k##S),        \
-        type_(InitListTensorType::Scalar) {}
+  // 2. A Tensor represented in `std::vector<ListInitTensor>` form, with value
+  //    `vec()`, Tensor scalar type `scalar_type()`, and Tensor sizes `sizes()`.
+  struct ListInitTensor {
+#define TENSOR(T, S)                                                                     \
+    ListInitTensor(T scalar) :                                                           \
+        scalar_(scalar), vec_(),                                                         \
+        sizes_(),                                                                        \
+        scalar_type_(at::k##S),                                                          \
+        type_(ListInitTensorType::Scalar) {}                                             \
+    ListInitTensor(std::vector<T> values) :                                              \
```

Comment:
Why do we need this constructor? `std::vector<T>` is implicitly convertible to `at::ArrayRef<T>`, so isn't it sufficient to just have that constructor?

Reply:
I tried removing the constructor that takes in `std::vector<T>`, but got the following error:
```
../aten/src/ATen/test/pow_test.cpp:225:39:   required from here
../aten/src/ATen/test/pow_test.cpp:111:36: error: no matching function for call to ‘tensor(const std::vector<int>&)’
   const auto tensor = torch::tensor(vals);
                                    ^
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../aten/src/ATen/test/pow_test.cpp:5:
../torch/csrc/autograd/generated/variable_factories.h:177:19: note: candidate: at::Tensor torch::tensor(torch::detail::ListInitTensor, const c10::TensorOptions&)
 inline at::Tensor tensor(detail::ListInitTensor list_init_tensor, const at::TensorOptions& options) {
                   ^
../torch/csrc/autograd/generated/variable_factories.h:177:19: note:   candidate expects 2 arguments, 1 provided
../torch/csrc/autograd/generated/variable_factories.h:181:19: note: candidate: at::Tensor torch::tensor(torch::detail::ListInitTensor)
 inline at::Tensor tensor(detail::ListInitTensor list_init_tensor) {
                   ^
../torch/csrc/autograd/generated/variable_factories.h:181:19: note:   no known conversion for argument 1 from ‘const std::vector<int>’ to ‘torch::detail::ListInitTensor’
```
I suspect the issue is that there might be too many conversions going on that the compiler just refuses to do it...

Reply:
Yikes. Could we make it take `const std::vector<T>& values` so that we don't incur an additional copy?

In addition, could we just call the ArrayRef ctor here? Something like
```
ListInitTensor(const std::vector<T>& values) : ListInitTensor(ArrayRef<T>(values)) {}`
```